README.md
# gaia esa API wrapper


[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)



## Installation

You can install the development version of gaia like so:

``` r
devtools::install_github("uskovgs/gaia")
```

## Example

This is a basic example which shows you how to solve a common problem:

## Cone search

``` r
library(gaia)
gaia_cone_search(
  ra = 43.1417,
  dec = 43.1674,
  radius = 40/3600, # 15 arcseconds
  table_name = "gaiadr3.gaia_source",
  cols = c("source_id", "ra", "ra_error", "dec", "dec_error"),
  verbose = TRUE)
#> SELECT TOP 50 source_id, ra, ra_error, dec, dec_error, DISTANCE(
#>     POINT('ICRS',ra, dec),
#>     POINT('ICRS', 43.1417, 43.1674)
#> ) AS dist
#> FROM
#>   gaiadr3.gaia_source
#> WHERE
#>   1 = CONTAINS(
#>   POINT('ICRS', ra, dec),
#>   CIRCLE('ICRS', 43.1417, 43.1674, 0.0111111111111111)
#> )
#> ORDER BY
#>   dist ASC
#> # A tibble: 4 × 6
#>   source_id    ra ra_error   dec dec_error   dist
#>     <int64> <dbl>    <dbl> <dbl>     <dbl>  <dbl>
#> 1      3e17  43.1  0.320    43.2    0.328  0.0101
#> 2      3e17  43.2  0.198    43.2    0.208  0.0102
#> 3      3e17  43.1  1.86     43.2    1.05   0.0108
#> 4      3e17  43.2  0.00955  43.2    0.0109 0.0110
```

## Searching by `source_id` column.

### `source_id` as type character

``` r
sources <- c("337387891762945152", "337384941121847552")
gaia_search_by_id(sources, 
                  cols = c("source_id", "ra", "ra_error", "dec", "dec_error"), 
                  verbose = TRUE)
#> SELECT  source_id, ra, ra_error, dec, dec_error
#> FROM gaiadr3.gaia_source
#> WHERE source_id IN (337387891762945152, 337384941121847552)
#> # A tibble: 2 × 5
#>   source_id    ra ra_error   dec dec_error
#>     <int64> <dbl>    <dbl> <dbl>     <dbl>
#> 1      3e17  43.2  0.00955  43.2    0.0109
#> 2      3e17  43.1  0.320    43.2    0.328
```

### Or using int64

``` r
# source_id column with long type (int64)
sources_int64 <- bit64::as.integer64(sources)
gaia_search_by_id(sources, 
                  cols = c("source_id", "ra", "ra_error", "dec", "dec_error"), 
                  verbose = TRUE)
#> SELECT  source_id, ra, ra_error, dec, dec_error
#> FROM gaiadr3.gaia_source
#> WHERE source_id IN (337387891762945152, 337384941121847552)
#> # A tibble: 2 × 5
#>   source_id    ra ra_error   dec dec_error
#>     <int64> <dbl>    <dbl> <dbl>     <dbl>
#> 1      3e17  43.2  0.00955  43.2    0.0109
#> 2      3e17  43.1  0.320    43.2    0.328
```


NEWS.md
# gaia 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


R/utils.R

check_table <- function(table_name) {
  table_name_list <- c("gaiaedr3.gaia_source",
                       "gaiadr3.gaia_source",
                       "gaiadr2.gaia_source",
                       "gaiadr1.gaia_source"
  )
  checkmate::assert_choice(table_name, choices = table_name_list)
}


check_format <- function(format) {
  format_list <- c("votable",
                   "votable_plain",
                   "csv",
                   "json",
                   "fits")
  checkmate::assert_choice(format, choices = format_list)
}

check_params <- function(ra, dec, r_arcmin, row_limit) {
  checkmate::assert_number(ra)
  checkmate::assert_number(dec)
  checkmate::assert_number(r_arcmin)
  checkmate::assert_int(row_limit)
}

check_colnames <- function(cols, table_name) {

  checkmate::assert_character(cols)

  if(table_name == "gaiaedr3.gaia_source")
    all_cols <- cols_availvable[["cols_edr3"]]
  if(table_name == "gaiadr3.gaia_source")
    all_cols <- cols_availvable[["cols_dr3"]]
  if(table_name == "gaiadr2.gaia_source")
    all_cols <- cols_availvable[["cols_dr1"]]
  if(table_name == "gaiadr1.gaia_source")
    all_cols <- cols_availvable[["cols_dr1"]]

  checkmate::assert_subset(cols, all_cols)
}


read_json_response <- function(resp) {

  cont <- resp %>%
    httr::content(as='text', encoding = "UTF-8") %>%
    jsonlite::fromJSON(simplifyDataFrame = TRUE,
                       bigint_as_char = TRUE)

  meta <- cont$metadata

  # cont$data is matrix
  data <- as.data.frame(cont$data)
  checkmate::assert_data_frame(data, min.rows = 1)

  colnames(data) <- meta$name

  convert_type_functions <- list(
    "long" = bit64::as.integer64,
    "char" = as.character,
    "double" = as.double,
    "float" = as.double,
    "int" = as.integer,
    "short" = as.integer,
    "boolean" = as.logical
  )
  col_types_functions <- convert_type_functions[meta$datatype]

  data1 <- purrr::map2_dfr(data, col_types_functions, ~.y(.x))

  return(data1)
}



R/gaia_cone.R
#' Make gaia query
#'
#' @param output_format output table format
#' @param adql ADQL string
#'
#' @return A `response()` object.
#' @export
#'
gaia_query_sync <- function(output_format, adql) {

  query = list(
    REQUEST = "doQuery",
    LANG    = "ADQL",
    FORMAT  = output_format,
    QUERY   = adql
  )

  httr::GET(
    url = "https://gea.esac.esa.int/tap-server/tap/sync",
    query = query
  )
}

#' Gaia cone search
#'
#' @param ra RA
#' @param dec DEC
#' @param table_name gaia table name (default = "gaiadr3.gaia_source")
#' @param row_limit number of the most nearest sources (default = 50). Set to *-1* for unlimited
#' @param verbose show ADQL query? (default = FALSE)
#' @param cols vector of column names (default = NULL)
#' @param radius circle radius in degrees
#'
#' @return A data frame
#' @export
#'
#' @examples
#' \dontrun{
#' gaia_cone_search(
#' ra = 95.1669125,
#' dec = 26.725528,
#' radius = 15/3600, # 15 arcseconds
#' table_name = "gaiadr3.gaia_source",
#' cols = c("source_id", "ra", "ra_error", "dec", "dec_error"),
#' verbose = TRUE)
#' }
gaia_cone_search <- function(ra, dec, radius,
                             row_limit = 50L,
                             table_name = "gaiadr3.gaia_source",
                             cols = NULL,
                             verbose = FALSE) {

  # gaia cone search work only for one point
  check_params(ra, dec, radius, row_limit)
  check_table(table_name)

  if(is.null(cols)) {
    cols <- "*"
  } else {
    check_colnames(cols, table_name)
    cols <- paste(cols, collapse = ", ")
  }

  row_limit <- ifelse(row_limit == -1, "", glue::glue("TOP {row_limit}"))

  adql <- glue::glue("
  SELECT {row_limit} {cols}, DISTANCE(
      POINT('ICRS',ra, dec),
      POINT('ICRS', {ra}, {dec})
  ) AS dist
  FROM
    {table_name}
  WHERE
    1 = CONTAINS(
    POINT('ICRS', ra, dec),
    CIRCLE('ICRS', {ra}, {dec}, {radius})
  )
  ORDER BY
    dist ASC
  ")

  if(verbose)
    print(adql)

  resp <- gaia_query_sync("json", adql) %>%
    read_json_response()

  return(resp)
}


#' Search gaia source by source_id
#'
#' @param id vector of source_id (bit64::integer64, character)
#' @param table_name gaia table name (default = "gaiadr3.gaia_source")
#' @param cols vector of column names (default = NULL)
#' @param verbose show ADQL query? (default = FALSE)
#'
#' @return A data frame
#' @export
#'
#' @examples
#' \dontrun{
#' sources <- c("3432266562766594688", "3432266562764508032")
#' sources_int64 <- bit64::as.integer64(sources)
#' gaia_search_by_id(sources, verbose = TRUE)
#' gaia_search_by_id(sources_int64, cols = c("source_id", "ra", "ra_error", "dec", "dec_error"))
#'}
gaia_search_by_id <- function(id,
                              table_name = "gaiadr3.gaia_source",
                              cols = NULL,
                              verbose = FALSE) {
  # checkmate::assert_int(id)
  checkmate::assert_multi_class(id, c("integer64", "character"))

  if(is.null(cols)) {
    cols <- "*"
  } else {
    check_colnames(cols, table_name)
    cols <- paste(cols, collapse = ", ")
  }

  ids <- id %>%
    paste(collapse = ", ") %>%
    paste0("(", ., ")")

  adql <- glue::glue("
  SELECT  {cols}
  FROM {table_name}
  WHERE source_id IN {ids}
  ")
#
#   adql <- glue::glue("
#   SELECT  {cols}
#   FROM {table_name}
#   WHERE source_id = {id}
#   ")

  if(verbose)
    print(adql)

  resp <- gaia_query_sync("json", adql)

  resp %>%
    read_json_response()

}



